x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformPersistentTasksExecutor.java

```diff
@@ -97,11 +100,69 @@ public TransformPersistentTasksExecutor(
             logger.debug(reason);
             return new PersistentTasksCustomMetaData.Assignment(null, reason);
         }
-        DiscoveryNode discoveryNode = selectLeastLoadedNode(
-            clusterState,
-            (node) -> node.isDataNode() && node.getVersion().onOrAfter(params.getVersion())
-        );
-        return discoveryNode == null ? NO_NODE_FOUND : new PersistentTasksCustomMetaData.Assignment(discoveryNode.getId(), "");
+        DiscoveryNode discoveryNode = selectLeastLoadedNode(clusterState, (node) -> nodeCanRunThisTransform(node, params, null));
+
+        if (discoveryNode == null) {
+            Map<String, String> explainWhyAssignmentFailed = new TreeMap<>();
+            for (DiscoveryNode node : clusterState.getNodes()) {
+                nodeCanRunThisTransform(node, params, explainWhyAssignmentFailed);
+            }
+            String reason = "Not starting transform ["
+                + params.getId()
+                + "], reasons ["
+                + explainWhyAssignmentFailed.entrySet().stream().map(e -> e.getKey() + ":" + e.getValue()).collect(Collectors.joining("|"))
+                + "]";
+
+            logger.debug(reason);
+            return new PersistentTasksCustomMetaData.Assignment(null, reason);
+        }
+
+        return new PersistentTasksCustomMetaData.Assignment(discoveryNode.getId(), "");
+    }
+
+    public static boolean nodeCanRunThisTransform(DiscoveryNode node, TransformTaskParams params, Map<String, String> explain) {
+        // data node?
+        if (node.isDataNode() == false) {
```

Comment:
I think this should only be checked for nodes that are on versions earlier than 7.7.

To match the documentation transforms should run on nodes that are defined as transform nodes, regardless of whether they are also data nodes.

And then since we didn't have transform nodes prior to 7.7 the BWC logic should be that pre-7.7 nodes are transform nodes if and only if they are data nodes.